torchao/prototype/smoothquant/core.py

```diff
@@ -0,0 +1,167 @@
+from dataclasses import dataclass
+from typing import Optional
+import torch
+import torch.nn.functional as F
+from torch.ao.quantization import PerChannelMinMaxObserver, MovingAverageMinMaxObserver
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    ZeroPointDomain,
+)
+from torchao.quantization.observer import (
+    AffineQuantizedObserverBase, PerRow
+)
+
+
+class SmoothQuantObserver(AffineQuantizedObserverBase):
+    def __init__(self,
+        weight: torch.Tensor,
+        alpha: float = 0.5,
+        quant_mode: str = "static", # or dynamic
+        n_calib_examples: int = 20,
+        quant_min: Optional[int] = None,
+        quant_max: Optional[int] = None,
+        eps: Optional[float] = None,
+        scale_dtype: Optional[torch.dtype] = None,
+        zero_point_dtype: Optional[torch.dtype] = None,
+        zero_point_domain = ZeroPointDomain.INT,
+    ):
+        """
+        A custom observer for SmoothQuant
+
+        Args:
+            weight: The weight tensor to be observed.
+            mapping_type: symmetric or asymmetric quantization of weight
+            alpha: The alpha value to determine smoothing factor
+            quant_mode: The mode of activation quantization, either static or dynamic
+            n_calib_examples: Number of examples used to calibrate observer
+            quant_min: The minimum quantized value
+            quant_max: The maximum quantized value
+            eps: The minimum scale to avoid dividing by zero.
+            scale_dtype: The data type of the scale tensor.
+            zero_point_dtype: The data type of the zero point tensor.
+            zero_point_domain: The domain of the zero point.
+        """
+        super().__init__(
+            MappingType.SYMMETRIC,
+            torch.int8,
+            PerRow(), 
+            quant_min = quant_min,
+            quant_max = quant_max,
+            eps = eps,
+            scale_dtype = scale_dtype,
+            zero_point_dtype = zero_point_dtype,
+            preserve_zero = True,
+            zero_point_domain = zero_point_domain,
+        )
+        assert weight.ndim == 2
+        self.weight = weight
+        self.n_calib_examples = n_calib_examples
+        self.inputs = []
+        self.device = self.weight.device
+        self.alpha = alpha
+        assert quant_mode in ["static", "dynamic"]
+        self.quant_mode = quant_mode
+        # act.shape = [mb, ic] (reshape if needed), wei.shape = [oc, ic]
+        # *_ic_obs are used to determine smoothing_factor
+        # *_mb/oc_obs are used to find qparams for quantization
+        self.act_ic_obs = PerChannelMinMaxObserver(
```

Comment:
nit: I think PerChannelMinMaxObserver can be replaced by our torchao AQT observer, see https://github.com/pytorch/ao/blob/f33cff7053530506d11085eb18e6b93f5e224faf/tutorials/calibration_flow/static_quant.py#L168 (need to set ch_axis correctly)

Reply:
Thanks for reviewing. Just curious. Why are we using observers in Torchao instead of those in PyTorch?

Reply:
PyTorch observers are not general enough since they only support per tensor, per channel etc. since now we also have per token, per group etc., observers in torchao are more general since we have the block_size arg that can represent all granularities. also pytorch observer has some args we want to get rid of like custom_qmin/qmax

Reply:
Thanks. I have refactored the observer part.